sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubJsonClient.java

```diff
@@ -185,6 +185,9 @@ public int publish(TopicPath topic, List<OutgoingMessage> outgoingMessages)
 
       // Payload.
       byte[] elementBytes = pubsubMessage.decodeData();
+      if (elementBytes == null) {
```

Comment:
Thanks! Could you add a unit test for this case?

Reply:
Returning `new byte[0]` instead of `null` is a change in the API. Existing users that have messages with no payload are probably already checking for null, and changing the returned value will break them.
I tried running the unit test you wrote without the changes to this file and didn't get an NPE, which means the Beam SDK doesn't trigger an NPE. (Technically if you try to call things like IncomingMessage.toString() you will get an NPE.)

So I'm sorry but we can't accept this change because it is backwards incompatible.

Reply:
The NPE is caused by this line: https://github.com/apache/beam/blob/4670344e5d9137020e128979a9154ee0794a5653/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSource.java#L818

My rationale for returning `new byte[0]` in PubsubJsonClient instead of checking from `null` at the site that causes the NPE in PubsubUnboundedSource is this is the behaviour of the PubsubGrpcClient, the other implementation of PubsubClient. Returning null in one implementation and an empty byte array in another makes the API for PubsubClient less clear. It also reduces the amount of null checking required in the codebase.

Do you consider this behaviour part of the API even though PubsubJsonClient is an implementation of PubsubClient which returns `new byte[0]` in a different implementation and PubsubJsonClient is only supposed to be used indirectly via PubsubUnboundedSource (via PubsubIO.read())? If so I will add null checks instead.

Reply:
You're right, I've created a FR to add this case to our integration test: https://issues.apache.org/jira/browse/BEAM-4589